### PR TITLE
BFD-1728 - Flyway script to fix new SNF claims table structure and migrate current SNF claims data

### PR DESCRIPTION
**Pull request by mackec1**
**JIRA Ticket:**
[BFD-1728](https://jira.cms.gov/browse/BFD-1728)

**User Story or Bug Summary:**
A prior flyway migration of the snf_claims and snf_claim_lines tables used incorrect db column data types; this PR is intended to correct that mistake and re-migrate the SNF claims data to the _new tables.

The initial discovery occurred during an ETL pipeline load in which the REV_CNTR_NDC_QTY failed to parse an incoming value of ".17"; the failure occurred because the column in the Excel spreadsheet had defined that column as an integer.

### What Does This PR Do?
This PR is intended to rectify errors in database table column definitions in which a values that may have decimal places are correctly handled. This is done by dropping the previous _snf_claims_new_ and _snf_claim_lines_new_ tables and then rebuilding.
 - snf_claim_lines_new:
    - change clm_id from varchar(15) to bigint
    - change line_num from numeric to smallint; this column is part of a primary key.
    - rev_cntr_ndc_qty, explicitly define length and scale (12,3); the scale is needed as the HMEM JDBC driver drops the scale unless specified.
  - snf_claims_new:
    - change clm_id from (varchar 15) to bigint.
    - change bene_id from (varchar 15) to bigint.
    - change clm_grp_id from numeric(12) to bigint.
- migrate data from snf_claims to snf_claims_new
  - add not null constraints 
  - add index to bene_id 
- migrate data from snf_claim_lines to snf_claim_lines_new
  - add not null constraints

### What Should Reviewers Watch For?
* verify that the db migration script correctly migrates the data and creates appropriate indexes and data constraints. 
* adheres to flyway script migration numbering and script filename.

**Note:** This flyway migration will run for several hours; this implies that no other builds should be deployed until such time that this flyway migration completes.

**Note:** This flyway migration will be followed by a corresponding code change PR to support BFD use of the "_new" table data  #(https://github.com/CMSgov/beneficiary-fhir-data/pull/1096)

### What Security Implications Does This PR Have?

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [X] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [X] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [X] No

    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [X] No

### What Needs to Be Merged and Deployed Before this PR?

Nothing unless another flyway script jumps the line and is deployed prior to this one; in that case, the V66__ will need to be modified to one more than the latest flyway script.

### Submitter Checklist

I have gone through and verified that...:

* [X] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [ X Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V66__new_schema_snf_claims_fix.sql [hunk 1]
```diff
@@ -0,0 +1,522 @@
...
+    alter column NCH_BLOOD_PNTS_FRNSHD_QTY    ${logic.alter-column-type} numeric;
+
+-- carrier_claim_lines table:
+--     REV_CNTR_UNIT_CNT            NUM  *   *
+--     REV_CNTR_NDC_QTY             NUM  *   *
+--
+alter table snf_claim_lines_new 
+    alter column REV_CNTR_UNIT_CNT  ${logic.alter-column-type} numeric;
+alter table snf_claim_lines_new     
+    alter column REV_CNTR_NDC_QTY   ${logic.alter-column-type} numeric;
```
Comment (christrotteradhoc) [inline apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V66__new_schema_snf_claims_fix.sql:42]: Looks like we are reverting all but clm_id and clm_line_num, right?

Given that, I think that it is more straightforward and easier to digest if we simply copy the V63 over to V66, make the fixes, and add a drop table at the beginning.

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V66__new_schema_snf_claims_fix.sql [hunk 2]
```diff
@@ -0,0 +1,522 @@
...
+                    prcdr_dt23,
+                    prcdr_dt24,
+                    prcdr_dt25 )
+select
+${logic.psql-only}  cast(clm_id as bigint),
+${logic.psql-only}  cast(bene_id as bigint),
+${logic.psql-only}  cast(clm_grp_id as bigint),  
+${logic.hsql-only}  convert(clm_id, SQL_BIGINT),
+${logic.hsql-only}  convert(bene_id, SQL_BIGINT),
+${logic.hsql-only}  convert(clm_grp_id, SQL_BIGINT),
```
Comment (brianburton) [inline apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V66__new_schema_snf_claims_fix.sql:270]: The psql/hsql for same columns are all psql then all hsql here but below they are psql/hsql for each column.  Might be better to follow the same pattern in both places.

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V66__new_schema_snf_claims_fix.sql [hunk 3]
```diff
@@ -0,0 +1,522 @@
+-- NEW_SCHEMA_SNF_CLAIMS_FIX.SQL
+-- flyway migration for SNF_CLAIMS_NEW and SNF_CLAIM_LINES_NEW tables into
+-- a new schema structure; previus flyway script, V63__nw_schema_snf_claims.sql
+-- whcich erroneously defined columns in the snf_claim_lines_new table
+-- as an integer. This script fixes that mistake and does a complete data
+-- migration of SNF claims data because the _new tables would have missed
+-- a weekend's ETL propcessing.
```
Comment (brianburton) [inline apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V66__new_schema_snf_claims_fix.sql:7]: There are some spelling errors in this comment that should be corrected.

**Comment by brianburton** [comment]
I didn't do a field by field comparison but I did confirm the converted columns match the spreadsheet and that none of the columns seem to be repeated.

Questions:

- What about the index created in V33.  Should it be dropped prior to the inserts and then re-added after they are complete?
- What about the `SET` lines from V33 for `max_parallel_workers` etc?  Are those no longer needed?

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V66__new_schema_snf_claims_fix.sql (hunk 1 above)
Comment (mackec1) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V66__new_schema_snf_claims_fix.sql]: agreed. dropped the tables an completely re-created

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V66__new_schema_snf_claims_fix.sql (hunk 2 above)
Comment (mackec1) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V66__new_schema_snf_claims_fix.sql]: yea, was torn between having all HSQL and PSQL or having the table columns in order; changed to all HSQl/PSQL together. thx.

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V66__new_schema_snf_claims_fix.sql (hunk 3 above)
Comment (mackec1) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V66__new_schema_snf_claims_fix.sql]: fixed. thx.

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V66__new_schema_snf_claims_fix.sql [hunk 4]
```diff
@@ -0,0 +1,821 @@
+-- NEW_SCHEMA_SNF_CLAIMS_FIX.SQL
+-- flyway migration for SNF_CLAIMS_NEW and SNF_CLAIM_LINES_NEW tables into
+-- a new schema structure; previous flyway script, V63__nw_schema_snf_claims.sql,
+-- which erroneously defined columns in the snf_claim_lines_new table
+-- as an integer. This script fixes that mistake and does a complete data
+-- migration of SNF claims data because the _new tables would have missed
+-- the prvious weekend's ETL propcessing.
```
Comment (christrotteradhoc) [inline apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V66__new_schema_snf_claims_fix.sql:7]: couple of spelling errors on this line

**Comment by christrotteradhoc** [comment]
looks good. I'm ready to approve once a couple of minor spelling errors in the description comment are resolved.

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V66__new_schema_snf_claims_fix.sql [hunk 5]
```diff
@@ -0,0 +1,821 @@
+-- NEW_SCHEMA_SNF_CLAIMS_FIX.SQL
+-- flyway migration for SNF_CLAIMS_NEW and SNF_CLAIM_LINES_NEW tables into
+-- a new schema structure; previous flyway script, V63__nw_schema_snf_claims.sql,
```
[the remark this answers is not in the document]
Comment (mackec1) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V66__new_schema_snf_claims_fix.sql]: fixed.

File: apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V66__new_schema_snf_claims_fix.sql (hunk 4 above)
Comment (mackec1) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V66__new_schema_snf_claims_fix.sql]: fixed.

Comment (christrotteradhoc) [reply apps/bfd-model/bfd-model-rif/src/main/resources/db/migration/V66__new_schema_snf_claims_fix.sql]: still have one left: 'propcessing'

**Review by christrotteradhoc — APPROVED** [review]

**Review by cbrunefearless — APPROVED** [review]
lgtm

**Pull request merged**